addOns/frontendscanner/src/main/java/org/zaproxy/zap/extension/frontendscanner/FrontEndScannerProxyListener.java

```diff
@@ -0,0 +1,114 @@
+import net.htmlparser.jericho.Element;
+import net.htmlparser.jericho.OutputDocument;
+import net.htmlparser.jericho.Source;
+import org.apache.log4j.Logger;
+import org.parosproxy.paros.core.proxy.ProxyListener;
+import org.parosproxy.paros.extension.history.ProxyListenerLog;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.api.API;
+
+/**
+ * The {@link ProxyListener} the {@link ExtensionFrontEndScanner} relies on.
+ */
+public class FrontEndScannerProxyListener implements ProxyListener {
+    private Logger LOGGER;
+
+    private FrontEndScannerAPI api;
+    private FrontEndScannerOptions options;
+
+    public FrontEndScannerProxyListener(FrontEndScannerAPI api, FrontEndScannerOptions options, Logger logger) {
+        this.api = api;
+        this.options = options;
+        this.LOGGER = logger;
+    }
+
+    @Override
+    public boolean onHttpRequestSend(HttpMessage msg) {
+        return true;
+    }
+
+    @Override
+    public boolean onHttpResponseReceive(HttpMessage msg) {
+        if (options.isEnabled() && msg.getResponseHeader().isHtml()) {
+            try {
+                String html = msg.getResponseBody().toString();
+
+                Source document = new Source(html);
+                List<Element> heads = document.getAllElements("head");
+                Element head = heads.isEmpty() ? null : heads.get(0);
+
+                if (head != null && msg.getHistoryRef() != null) {
+                    String host = msg.getRequestHeader().getHeader("host");
+                    String frontEndApiUrl =
+                            API.getInstance().getCallBackUrl(this.api, "https://" + host);
+
+                    int historyReferenceId = msg.getHistoryRef().getHistoryId();
+
+                    StringBuilder injectedContentBuilder =
+                            new StringBuilder(200)
+                                    .append("<script src='")
+                                    .append(frontEndApiUrl)
+                                    .append("?action=getFile")
+                                    .append("&filename=front-end-scanner.js")
+                                    .append("&historyReferenceId=")
+                                    .append(historyReferenceId)
+                                    .append("'></script>");
+
+                    String injectedContent = injectedContentBuilder.toString();
+
+                    OutputDocument newResponseBody = new OutputDocument(document);
+                    int insertPosition = head.getChildElements().get(0).getBegin();
+                    newResponseBody.insert(insertPosition, injectedContent);
+
+                    msg.getResponseBody().setBody(newResponseBody.toString());
+
+                    int newLength = msg.getResponseBody().length();
+                    msg.getResponseHeader().setContentLength(newLength);
+                } else {
+                    LOGGER.debug("<head></head> is missing in the response");
```

Comment:
Should a head be injected before the body start tag if one doesn't exist or would that break other things?

Reply:
I don't think it would... I will give it a try.

As this code was already written (most of the content from `FrontEndScannerProxyListener` has been copy-pasted from `ExtensionFrontEndScanner`), and no issues were mentioning that, I did not touch it.

But more generally, it could be taking these two cases into consideration:
  1. no `<head>` present;
  2. empty `<head>`.